### PR TITLE
SWATCH-2571: Implement billable usage "gratis" status

### PR DESCRIPTION
**Pull request by Sgitario**
Jira issue: SWATCH-2571

## Description
- support enableGratisUsage in the metrics definition
- set "enableGratisUsage" to true only for ansible and the metric "Managed-nodes"
- save the gratis remittance records if any applicable contract has the gratis flag enabled

## Testing
1. podman-compose up
2. setup database schema: `./gradlew liquibaseUpdate`
3. insert test data:

What is important here is that:
- the contract has the product 'ansible-aap-managed': because it's the only product with the gratis flag enabled
- the contract start date starts after the current month (when writing these steps is '2024-08-01')

```
insert into offering(sku) values ('MW02393');
insert into sku_product_tag(sku,product_tag) values ('MW02393','ansible-aap-managed');
INSERT INTO contracts VALUES ('b3a5e8b7-6260-48f4-a9eb-63a584763ce9', '12914277', now(), '2024-08-01 01:01:01.717275+00', NULL, 'org123', 'MW02393', 'aws', 'aws123', 'ansible-aap-managed', 'bcwuzc8ww10vvxfair55mliy8');
```
4. start contract service: `SERVER_PORT=8001 QUARKUS_MANAGEMENT_PORT=9001 ./gradlew :swatch-contracts:quarkusDev`
5. start billable usage service: `./gradlew :swatch-billable-usage:quarkusDev`
6. create message in `platform.rhsm-subscriptions.tally`:

What is important here is that:
- the usage is within the above contract time window (starts after 2024-08-01).

```
{"org_id": "org123", "tally_snapshots": [{"billing_provider":"aws", "billing_account_id": "aws123", "snapshot_date": "2024-08-10T01:10:28Z", "product_id": "ansible-aap-managed", "sla": "Premium", "usage": "Production", "granularity": "Hourly", "tally_measurements": [{"hardware_measurement_type": "AWS", "metric_id": "Managed-nodes", "value": 12, "currentTotal": 100}]}]}
``` 

7. verification: 

The billable remittance should be created with gratis status:

```
select * from billable_usage_remittance where billing_account_id='aws123'
```

There should be only one with status "gratis".

And there should not be any new messages in the topic "platform.rhsm-subscriptions.billable-usage" (note that this topic is created only when any of the swatch-producer-aws or swatch-producer-azure is started).

### CODE REVIEW

File: swatch-billable-usage/src/main/java/com/redhat/swatch/billable/usage/services/BillableUsageService.java
Comment (kahowell) [inline swatch-billable-usage/src/main/java/com/redhat/swatch/billable/usage/services/BillableUsageService.java:216]: Rather than repeat a bunch of logic from `createRemittance`, it'd be better to add a parameter to createRemittance for the `RemittanceStatus value`.

We _do_ want the value to be populated using the same logic:
```java
    // Remitted value should be set to usages metric_value rather than billing_value
    newRemittance.setRemittedPendingValue(usageCalc.getRemittedValue());
    newRemittance.setRemittancePendingDate(usageCalc.getRemittanceDate());
```.

So the only difference should be in the `status` value.

File: swatch-billable-usage/src/main/java/com/redhat/swatch/billable/usage/services/ContractsController.java
Comment (kahowell) [inline swatch-billable-usage/src/main/java/com/redhat/swatch/billable/usage/services/ContractsController.java:111]: As mentioned elsewhere, rather than short-circuiting, the coverage should be considered gratis only when all matching contracts are gratis. So it's probably easiest to store vars for `hasGratisContracts` and `hasNonGratisContracts` and then set `coverage.setGratis(hasGratisContracts && !hasNonGratisContracts)` outside of the loop (or something equivalent).

File: swatch-billable-usage/src/main/java/com/redhat/swatch/billable/usage/services/ContractsController.java
Comment (kahowell) [inline swatch-billable-usage/src/main/java/com/redhat/swatch/billable/usage/services/ContractsController.java:133]: We simplified the logic during refinement, the only condition now is that the startDate is after the start of the current month:
```suggestion
    OffsetDateTime startDate = contract.getStartDate();
    return startDate != null && startDate.isAfter(clock.startOfCurrentMonth());
```

**Review by kahowell — CHANGES_REQUESTED** [review]

File: swatch-billable-usage/src/main/java/com/redhat/swatch/billable/usage/services/BillableUsageService.java
Comment (Sgitario) [reply swatch-billable-usage/src/main/java/com/redhat/swatch/billable/usage/services/BillableUsageService.java:165]: I got confused because we're using this amount in the method `calculateBillableUsage` to calculate the actual usage calculation... But I'm reverting this change.

File: swatch-billable-usage/src/main/java/com/redhat/swatch/billable/usage/services/ContractsController.java
Comment (Sgitario) [reply swatch-billable-usage/src/main/java/com/redhat/swatch/billable/usage/services/ContractsController.java:111]: thanks for clarifying, will change it

File: swatch-billable-usage/src/main/java/com/redhat/swatch/billable/usage/services/ContractsController.java
Comment (Sgitario) [reply swatch-billable-usage/src/main/java/com/redhat/swatch/billable/usage/services/ContractsController.java:133]: Thanks for clarifying. Will change it.

**Review by kartikshahc — APPROVED** [review]
Haven't tested the changes yet but, the code LGTM after the feedback from Kevin. Will test the code today.

File: swatch-product-configuration/src/main/resources/subscription_configs/Ansible/ansible-aap-managed.yaml
Comment (kahowell) [inline swatch-product-configuration/src/main/resources/subscription_configs/Ansible/ansible-aap-managed.yaml:29]: This should be set on `Managed-nodes` rather than `Instance-hours`

**Review by kahowell — CHANGES_REQUESTED** [review]
One tiny tweak needed to config, otherwise LGTM

File: swatch-product-configuration/src/main/resources/subscription_configs/Ansible/ansible-aap-managed.yaml
Comment (kartikshahc) [reply swatch-product-configuration/src/main/resources/subscription_configs/Ansible/ansible-aap-managed.yaml:29]: Interesting! Agreed that's what I thought I had it before when I merged the main branch to this. :)

**Comment by kartikshahc** [comment]
> Haven't tested the changes yet but, the code LGTM after the feedback from Kevin. Will test the code today.

I later thought we will test things in https://issues.redhat.com/browse/SWATCH-2755 which we will still do. I am going to proactively test this tomorrow after speaking with Barnaby to discover any issues upfront.

File: swatch-product-configuration/src/main/resources/subscription_configs/Ansible/ansible-aap-managed.yaml
Comment (Sgitario) [reply swatch-product-configuration/src/main/resources/subscription_configs/Ansible/ansible-aap-managed.yaml:29]: Thanks for spotting! Yes, this was messed up after rebasing to merge :(

**Comment by Sgitario** [comment]
PR updated by fixing the metric configuration that uses the enableGratis flag.

**Comment by diegomaranhao** [comment]
@Sgitario below are the reproducer steps with the Iqe method

1. Deploy the EE and Iqe pod
`oc rsh <Iqe pod>`
Open Iqe shell
`ENV_FOR_DYNACONF=clowder_smoke iqe shell`

`application=app`

2. Importing 
```
import random
import datetime
from iqe_rhsm_subscriptions.rhsmlib import config_parser
import fauxfactory
```

3. Create Ansible Contract
```
billing_account_id = "test" + fauxfactory.gen_alpha(3).lower()
billing_provider = "aws"
product_id = "ansible-aap-managed"
metric = "Managed-nodes"
contract_data = {  
    "instance_hours": random.randint(0, 10),  
    "manage_cores": random.randint(0, 7),  
    "billing_provider": billing_provider,  
    "billing_account_id": billing_account_id,  
    "start_date": start_date,
    "end_date": end_date,
} 

contract = app.rhsm_subscriptions.create_contract(  
    product_id=product_id,  
    **contract_data,  
)

Out[12]: 
[{'uuid': 'a6f2f1ba-f87e-4a4c-b59d-1b28c15d505c',
  'subscription_number': '1542938',
  'sku': 'MCT4701',
  'start_date': '2024-09-01T14:53:59Z',
  'end_date': '2024-10-12T14:53:59Z',
  'org_id': '3340851',
  'billing_provider': 'aws',
  'billing_provider_id': '5f8bfa10-faad-478b-8432-25944e285d5f;aws_customer_id_123;1234567',
  'billing_account_id': 'testthm',
  'product_tags': ['ansible-aap-managed'],
  'vendor_product_code': '5f8bfa10-faad-478b-8432-25944e285d5f',
  'metrics': []}]
```

4. Creating mock subscription.
```
app.rhsm_subscriptions.create_mock_subscription(  
    product_id=product_id,  
    billing_provider=billing_provider,  
    billing_account_id=billing_account_id,  
    start_day=(datetime.date.today() - datetime.timedelta(days=10)).strftime("%Y-%m-%d"),  
)
Out[13]: 
[{'id': 639214,
  'subscriptionNumber': 2241942,
  'subscriptionProducts': [{'sku': 'MCT4701'}],
  'webCustomerId': '3340851',
  'quantity': 1,
  'oracleAccountNumber': '',
  'effectiveStartDate': 1722644053415.0,
  'effectiveEndDate': 1755044053415.0,
  'externalReferences': {'awsMarketplace': {'accountID': None,
    'subscriptionID': None,
    'customerID': 'testthm',
    'productCode': 'testProductCode',
    'sellerAccount': '1234567',
    'customerAccountID': 'testthm'}}}]

```

5. Add an Ansible event with billing_provider=aws, billing account_id=testthm

```
value = random.randint(1, 20)  
hours_value = random.randint(1, 20)  
data = {  
    metric.lower(): value,  
    "billing_provider": billing_provider,  
    "billing_account_id": billing_account_id,  
    "hours": hours_value,  
    "event_source": "cost-management",  
}  

host_instance = application.rhsm_subscriptions.create_mock_payg_cluster(  
	product_id=product_id, **data  
) 

2024-08-13 15:07:08.148 [    INFO] [root] ansible-aap-managed PAYG Cluster 40239a9c-f03e-4ad7-a900-3a66493eb43e created successfully!!
```

Sync hourly tally 
```
application.rhsm_subscriptions.sync_tally_hourly(  
    product_id=product_id,  
    metric_id=metric,  
    perform_metering=False,  
    wait=True,  
    wait_for_remittance=True,  
)
```

Check remittance for Ansible with `billing_account_id=testthm`
```
application.rhsm_subscriptions.get_account_remittance(  
	product_id=product_id,  
	metricId=metric,  
	billingProvider=billing_provider,  
	billingAccountId=billing_account_id,
)

Out[17]: 
[{'orgId': '3340851',
  'productId': 'ansible-aap-managed',
  'metricId': 'Managed-nodes',
  'billingProvider': 'aws',
  'billingAccountId': 'testthm',
  'remittedValue': 0.0,
  'accumulationPeriod': None,
  'remittanceDate': None,
  'remittanceStatus': None}]

```

Repeat step 5 and create one more event with the same details
```
application.rhsm_subscriptions.create_mock_payg_cluster(  
	product_id=product_id, **data  
) 

2024-08-13 15:25:07.156 [    INFO] [root] ansible-aap-managed PAYG Cluster 70eaead9-1d93-4929-a12b-77347c0f0ad0 created successfully!!
```

Sync hourly tally again 
```
application.rhsm_subscriptions.sync_tally_hourly(  
    product_id=product_id,  
    metric_id=metric,  
    perform_metering=False,  
    wait=True,  
    wait_for_remittance=True,  
)
```

Check remittance for Ansible and with `billing_account_id=testthm`
```
application.rhsm_subscriptions.get_account_remittance(  
	product_id=product_id,  
	metricId=metric,  
	billingProvider=billing_provider,  
	billingAccountId=billing_account_id,
)

Out[19]: 
[{'orgId': '3340851',
  'productId': 'ansible-aap-managed',
  'metricId': 'Managed-nodes',
  'billingProvider': 'aws',
  'billingAccountId': 'testthm',
  'remittedValue': 0.0,
  'accumulationPeriod': None,
  'remittanceDate': None,
  'remittanceStatus': None}]
```

Add one more event with a different billing_account_id=test-150
```
data = {  
    metric.lower(): value,  
    "billing_provider": billing_provider,  
    "billing_account_id": "test-150",  
    "hours": hours_value,  
    "event_source": "cost-management",  
} 

application.rhsm_subscriptions.create_mock_payg_cluster(  
	product_id=product_id, **data  
)

```

Sync hourly tally again
```
application.rhsm_subscriptions.sync_tally_hourly(  
    product_id=product_id,  
    metric_id=metric,  
    perform_metering=False,  
    wait=True,  
    wait_for_remittance=True,  
)
```

Get account remittance
```
application.rhsm_subscriptions.get_account_remittance(
  product_id=product_id,
)
2024-08-13 15:32:45.558 [    INFO] [root] GET request response url: http://swatch-billable-usage-service:8000/api/swatch-billable-usage/internal/remittance/accountRemittances?productId=ansible-aap-managed&orgId=3340851, status_code: 200
2024-08-13 15:32:45.558 [    INFO] [root] Response: [
    {
        "accumulationPeriod": null,
        "billingAccountId": null,
        "billingProvider": null,
        "metricId": null,
        "orgId": "3340851",
        "productId": "ansible-aap-managed",
        "remittanceDate": null,
        "remittanceStatus": null,
        "remittedValue": 0.0
    }
]
Out[22]: 
[{'orgId': '3340851',
  'productId': 'ansible-aap-managed',
  'metricId': None,
  'billingProvider': None,
  'billingAccountId': None,
  'remittedValue': 0.0,
  'accumulationPeriod': None,
  'remittanceDate': None,
  'remittanceStatus': None}]
```
Consulting tally db
```
swatch-tally-db=# select * from billable_usage_remittance;
 account_number | org_id | product_id | metric_id | accumulation_period | sla | usage | billing_provider | billing_account_id | remitted_pending_value | remittance_pending_date | retry_after | tally_id | hardware_measurement_type | uuid | billed_on | error_code | status 
----------------+--------+------------+-----------+---------------------+-----+-------+------------------+--------------------+------------------------+-------------------------+-------------+----------+---------------------------+------+-----------+------------+--------
(0 rows)


```
At this point I was expecting that the remittance was different than `None` and the entry in billable_usage_remittance database table with status = "gratis"

**Comment by kartikshahc** [comment]
Haven't checked the steps provided by @diegomaranhao. But, @Sgitario I think I found couple of bugs after some tedious manual testing. I still have to do some edge case testing but, below are my notes.
1) We have to check with @kahowell and then Ansible folks if we require billing_account_id and billing_provider in event. Because without that the contract won't match. Currently, it seems they don't have it configured.

2) 
Event from ansible looks like below without above two fields but, you can also do some splunk check to find the latest events:
`{"sla": "Premium", "usage": "Production", "org_id": "org123", "event_id": "5e02b8a2-39e7-4371-9fc5-8a6f01f83a25", "timestamp": "2024-08-06T13:00:00Z", "conversion": false, "event_type": "managed-host", "expiration": "2024-08-06T14:00:00Z", "instance_id": "event_kafka_50c87c49-d789-4751-983e-2adafe75c06f/NiWexktbeV", "product_tag": ["ansible-aap-managed"], "record_date": "2024-08-14T01:12:12.502670052Z", "display_name": "automation__cluster_event_kafka_50c87c49-d789-4751-983e-2adafe75c06f", "event_source": "urn:redhat:source:console:app:aap-controller-billing", "measurements": [{"value": 2.0, "metric_id": "Managed-nodes"}], "service_type": "Ansible Managed Node", "billing_provider": "aws", "billing_account_id": "aws123"}`

Run hourly tally and below tally snapshot is created:
```
{
  "org_id": "org123",
  "tally_snapshots": [
    {
      "id": "e7cfe013-d0e7-4e03-b411-af2a9d9c303a",
      "billing_provider": "aws",
      "billing_account_id": "aws123",
      "snapshot_date": "2024-08-06T13:00:00Z",
      "product_id": "ansible-aap-managed",
      "sla": "Premium",
      "usage": "Production",
      "granularity": "Hourly",
      "tally_measurements": [
        {
          "hardware_measurement_type": "PHYSICAL",
          "metric_id": "MANAGED_NODES",
          "value": 2.0,
          "currentTotal": 2.0
        }
      ]
    }
  ]
}
```

Since we change metric_id to uppercase MANAGED_NODES we have to do something like below since it wasn't seeing it as a gratis contract product.
```
boolean isGratisContract =
        SubscriptionDefinition.isMetricGratis(usage.getProductId(), MetricId.fromString(usage.getMetricId()).toString());
```

**Review by kartikshahc — CHANGES_REQUESTED** [review]
Manual test failed with above comments

**Comment by Sgitario** [comment]
@diegomaranhao I followed the steps you wrote (I fixed the contract creation to not use any start_date and end_date, and also add the opt_in call before tallying). 

After fixing the issue that @kartikshahc found about the metric ID, now the usage status is `gratis` when you do:

- Sync hourly tally again
```
application.rhsm_subscriptions.sync_tally_hourly(  
    product_id=product_id,  
    metric_id=metric,  
    perform_metering=False,  
    wait=True,  
    wait_for_remittance=True,  
)
```

Check remittance for Ansible and with billing_account_id=testthm
```
application.rhsm_subscriptions.get_account_remittance(  
	product_id=product_id,  
	metricId=metric,  
	billingProvider=billing_provider,  
	billingAccountId=billing_account_id,
)
Out[19]: 
[{'orgId': '3340851',
  'productId': 'ansible-aap-managed',
  'metricId': 'Managed-nodes',
  'billingProvider': 'aws',
  'billingAccountId': 'testdov',
  'remittedValue': 44.0,
  'accumulationPeriod': '2024-08',
  'remittanceDate': '2024-08-14T08:11:50.937151Z',
  'remittanceStatus': 'gratis'}]
```

Btw, I didn't need to create any additional event to see it. This means that I didn't do any steps beyond the `Add one more event with a different billing_account_id=test-150` statement.

**Review by kahowell — APPROVED** [review]

**Comment by kahowell** [comment]
> We have to check with @kahowell and then Ansible folks if we require billing_account_id and billing_provider in event. Because without that the contract won't match. Currently, it seems they don't have it configured.

Yeah, this is an oversight. We'll need to makes sure they send us those attributes.

**Review by diegomaranhao — APPROVED** [review]

**Review by kartikshahc — CHANGES_REQUESTED** [review]
@Sgitario Found another bug.
If the contract begins in the current month but doesn't end until the following year, and if events are coming for the second month, they are still getting considered as gratis.
I think you have to consider usage object in the below check. Like if the snapshot date is in next month then send the usage.
```
  private boolean isContractCompatibleWithGratis(Contract contract, BillableUsage usage) {
    OffsetDateTime startDate = contract.getStartDate();
    return startDate != null && startDate.isAfter(clock.startOfCurrentMonth()) && usage.getSnapshotDate().isBefore(endOfCurrentMonth());
  }
```

Second event I tried:
`{"sla": "Premium", "usage": "Production", "org_id": "org123", "event_id": "7e02b8a2-39e7-4371-9fc5-8a6f01f83a25", "timestamp": "2024-09-02T11:00:00Z", "conversion": false, "event_type": "managed-host", "expiration": "2024-09-06T14:00:00Z", "instance_id": "event_kafka_50c87c49-d789-4751-983e-2adafe75c06f/NiWexktbeV", "product_tag": ["ansible-aap-managed"], "record_date": "2024-09-18T18:29:55.854681891Z", "display_name": "automation__cluster_event_kafka_50c87c49-d789-4751-983e-2adafe75c06f", "event_source": "urn:redhat:source:console:app:aap-controller-billing", "measurements": [{"value": 4.0, "metric_id": "Managed-nodes"}], "service_type": "Ansible Managed Node", "billing_provider": "aws", "billing_account_id": "aws123"}`

**Comment by kartikshahc** [comment]
> @Sgitario Found another bug. If the contract begins in the current month but doesn't end until the following year, and if events are coming for the second month, they are still getting considered as gratis. I think you have to consider usage object in the below check. Like if the snapshot date is in next month then send the usage.
> 
> ```
>   private boolean isContractCompatibleWithGratis(Contract contract, BillableUsage usage) {
>     OffsetDateTime startDate = contract.getStartDate();
>     return startDate != null && startDate.isAfter(clock.startOfCurrentMonth()) && usage.getSnapshotDate().isBefore(endOfCurrentMonth());
>   }
> ```
> 
> Second event I tried: `{"sla": "Premium", "usage": "Production", "org_id": "org123", "event_id": "7e02b8a2-39e7-4371-9fc5-8a6f01f83a25", "timestamp": "2024-09-02T11:00:00Z", "conversion": false, "event_type": "managed-host", "expiration": "2024-09-06T14:00:00Z", "instance_id": "event_kafka_50c87c49-d789-4751-983e-2adafe75c06f/NiWexktbeV", "product_tag": ["ansible-aap-managed"], "record_date": "2024-09-18T18:29:55.854681891Z", "display_name": "automation__cluster_event_kafka_50c87c49-d789-4751-983e-2adafe75c06f", "event_source": "urn:redhat:source:console:app:aap-controller-billing", "measurements": [{"value": 4.0, "metric_id": "Managed-nodes"}], "service_type": "Ansible Managed Node", "billing_provider": "aws", "billing_account_id": "aws123"}`

Actually after thinking a bit, I believe this may not be required since the after condition will be false for next month when we are in next month. Approving the PR since it LGTM

**Pull request merged**
